File: backend/src/models/Flag.js
const mongoose = require('mongoose');

const flagSchema = new mongoose.Schema({
  type: {
    type: String,
    required: true,
  },
  adminId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Admin',
  },
  studentId: {
    type: String,
    trim: true,
    uppercase: true,
  },
  details: String,
  sessionId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Session',
  },
  timestamp: {
    type: Date,
    default: Date.now,
  },
  resolved: {
    type: Boolean,
    default: false,
  },
  resolvedBy: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Admin',
  },
  resolvedAt: Date,
});

flagSchema.index({ type: 1, timestamp: -1 });
flagSchema.index({ adminId: 1, timestamp: -1 });
flagSchema.index({ resolved: 1, timestamp: -1 });

module.exports = mongoose.model('Flag', flagSchema);

File: backend-rust/src/models/flag.rs
use chrono::{DateTime, Utc};
use mongodb::bson::oid::ObjectId;
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Flag {
    #[serde(rename = "_id", skip_serializing_if = "Option::is_none")]
    pub id: Option<ObjectId>,
    pub flag_type: String,
    pub admin_id: Option<ObjectId>,
    pub student_id: Option<ObjectId>,
    pub details: Option<String>,
    pub session_id: Option<ObjectId>,
    #[serde(with = "bson::serde_helpers::datetime::FromChrono04DateTime")]
    pub timestamp: DateTime<Utc>,
    pub resolved: bool,
    pub resolved_by: Option<ObjectId>,
    #[serde(default, with = "crate::models::optional_chrono_bson")]
    pub resolved_at: Option<DateTime<Utc>>,
}

impl Flag {
    pub fn collection_name() -> &'static str {
        "flags"
    }
}
